### ISSUE TITLE
Add "Security" section

### ISSUE THREAD

**Issue by kamilmysliwiec** [nestjs/docs.nestjs.com#109]
## I'm submitting a...

<pre><code>
[ ] Regression 
[ ] Bug report
[ ] Feature request
[x] Documentation issue or request (new chapter/page)
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Expected behavior

As a user, I would like to see Security chapter.

## Environment

<pre><code>
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>

**Comment by ahrnee** [comment]
A comprehensive security section would be extremely helpful.  There is a lot of complexity and confusion around the best way to accomplish tight security using NestJS - which is essential if it is to be used for production applications.  There are very few examples available to follow.  The examples that do exist all seem to either be too simple, have problems, or be somewhat contradictory.

**For example**:
 - [NesJS Basic Auth and Sessions Blogpost](https://blog.exceptionfound.com/index.php/2018/06/07/nestjs-basic-auth-and-sessions/) (artonio)
   - Overly simple implementation for a production system, and still ends up being a very complex example
   - Even with this simple example, finds (and reports) a bug in the Nest passport library, which is as-yet unfixed
   - No CRSF protection
   - No use of helmet or such for additional security

 - [NestJS Auth Chapter](https://docs.nestjs.com/techniques/authentication)
   - I have found the examples/docs not in depth enough to provide a solid foundation for production use
   - Not clear how this would integrate with social login, XRSF etc

 - [NestJS Starter Project](https://github.com/CanKattwinkel/starter) (CanKattwinkel)
   - Seems to use NestJS best practices (like using Nest middleware)
   - Gets into a lot of custom complexity, with sessions and XSRF protection, which makes the auth specific code hard to follow and identify
   - Not leveraging the Nest/Passport libraries
   - No license, so can be used for example only

 - [Nest-Angular open Source Project](https://github.com/bojidaryovchev/nest-angular) (bojidaryovchev)
   - Comprehensive example including social authentication, and helmet
   - Seems to not use standard NestJS techniques (uses express and express middleware directly instead of NestJS style, uses custom JOI validation via middleware)
   - No CRSF protection
  
**Wishlist**
  - How to implement local authentication using JWT Bearer Tokens
    - How to accommodate the token timeout and refresh scenarios
    - How to position this to accommodate both web and mobile clients
 - How to integrate auth via social login
 - How to implement CRSF protection
 - How to implement additional hardening (such as via helmet)
 - Nice to have: 2FA example

Any guidance for the above would be much appreciated.  I think that NestJS is a fantastic project, and would love to start using it in production with confidence that I have got the security right.

Thanks for all the time and effort on NestJS!

**Comment by bhaidar** [comment]
Any follow up on how to use **CSRF** in Nestjs?

**Comment by johnbiundo** [comment]
@bhaidar I think most of the considerations are not specific to Nest.  I use the csurf module, and cookie parser.  I also use express sessions for session management, and found that there's a side-effect in relying only on sessions for storing the csurf secret (probably particular to my use case), hence relying on cookies instead for that function.  So my main.ts looks like

```typesript
...
  app.use(cookieParser());
  app.use(csurf({ cookie: true }));
...
```

I mentioned the details above because they are generic Express/Sessions/CSurf issues, not Nest-specific, and illustrate the point that if your use case is complex (e.g., also using sessions), you may have to do some additional research into those packages.

Another aspect of the CSRF equation is your client side.  In my case, for example, I have a SPA and use Angular, which handles the client side transparently.  If you have an MVC style app, you'll need to do more work in how you render pages so that you pass the CSRF token to the forms so they're passed back on POSTs (just as one example).

**Comment by bhaidar** [comment]
Thanks @johnbiundo 
I figured out that CSRF should be used with MVC apps rather than SPAs. I also use Angular with Nest.

I believe the docs need more work regarding MVC unless building MVC apps with Nest or Node is not popular.

**Comment by kamilmysliwiec** [comment]
MVC apps are very rare nowadays @bhaidar
